flecsi/exec/local/bind_prologue.hh

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2016, Triad National Security, LLC
+// All rights reserved.
+
+#ifndef FLECSI_EXEC_LOCAL_BIND_PROLOGUE_HH
+#define FLECSI_EXEC_LOCAL_BIND_PROLOGUE_HH
+
+#include <flecsi-config.h>
+
+#include "flecsi/data/privilege.hh"
+#include "flecsi/exec/local/policy.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/annotation.hh"
+#include "flecsi/util/demangle.hh"
+
+#include <functional>
+#include <tuple>
+
+namespace flecsi {
+inline flog::devel_tag bind_prologue_tag("bind_prologue");
+}
+
+// task_prologue is implemented per backend:
+#if FLECSI_BACKEND == FLECSI_BACKEND_legion
+#include "flecsi/exec/leg/bind_accessors.hh"
+#elif FLECSI_BACKEND == FLECSI_BACKEND_mpi
+#error "The MPI backend has no need for bind_accessors"
+#elif FLECSI_BACKEND == FLECSI_BACKEND_hpx
+#include "flecsi/exec/hpx/bind_accessors.hh"
+#endif
+
+namespace flecsi {
+namespace exec {
+
+#ifdef DOXYGEN // implemented per-backend
+/// Handling for low-level special task parameters/arguments.
+/// The exact member function signatures may vary between backends.
+template<task_processor_type_t Proc>
+struct bind_accessor {
+protected:
+  /// Constructible from backend-specific arguments.
+  template <typename...Args>
+  bind_accessor(Args&&... args);
+
+  /// Send a raw field reference to a raw accessor.
+  template<typename T, Privileges P>
+  void visit(data::accessor<data::raw, T, P> &);
+};
+#endif
+
+template<task_processor_type_t ProcessorType>
+struct bind_prologue : bind_accessors<ProcessorType> {
+
+  template<class A, class... Args>
+  bind_prologue(A & a, Args &&... args)
+    : bind_accessors<ProcessorType>(std::forward<Args>(args)...) {
+    util::annotation::rguard<util::annotation::execute_bind_prolog> ann;
+    std::apply([&](auto &... aa) { (visit(aa), ...); }, a);
```

Comment:
While traversing the arguments, the `bind_accessors` base class (at least for HPX) also collects the reduction operations that need to be performed after the actual task is executed. That means that the object has to stay alive for the duration of the task being run.

Also, FWIW, I followed the scheme that pre-existed for the `task_prologue`, where the constructor is performing the traversal as well.

Reply:
Sorry: I didn't remember the usage in `apply_delayed_prolog` when I wrote this.  It's fine that it's a class, then, although I don't know why it can't continue to use `operator()` for the scan.

For `task_prologue`, the object is useful after that initialization because we need to access the data it built during the constructor call.  Here, there was (in the Legion backend, it turns out!) no reason to ever keep the `bind_prologue` object, so that it looked and acted like a function call without being one.

Reply:
Combining the `operator()()` with the constructor is pure convenience. I can certainly revert this change, if you prefer having it separately.

Reply:
I think it's conceptually cleaner to have the `operator()` separate, since it lets us describe it as "collect the field-pointer information" and "install it in the accessors".  That said, if you think it's more convenient this way I don't object, since (for the HPX backend) we do have to have the object around after its constructor anyway.